setup.py

```diff
@@ -1,73 +1,110 @@
+        # To support Python 2, we have to avoid super(), since distutils is all
+        # old-style classes.
+        build_ext.run(self)
 
     def build_extension(self, ext):
         extdir = os.path.abspath(os.path.dirname(self.get_ext_fullpath(ext.name)))
+
         # required for auto-detection of auxiliary "native" libs
         if not extdir.endswith(os.path.sep):
             extdir += os.path.sep
 
-        cmake_args = ['-DCMAKE_LIBRARY_OUTPUT_DIRECTORY=' + extdir,
-                      '-DPYTHON_EXECUTABLE=' + sys.executable]
+        cfg = "Debug" if self.debug else "Release"
 
-        cfg = 'Debug' if self.debug else 'Release'
-        build_args = ['--config', cfg]
+        # Set Python_EXECUTABLE instead if you use PYBIND11_FINDPYTHON
+        cmake_args = [
+            "-DCMAKE_LIBRARY_OUTPUT_DIRECTORY={}".format(extdir),
+            "-DPYTHON_EXECUTABLE={}".format(sys.executable),
+            "-DEXAMPLE_VERSION_INFO={}".format(self.distribution.get_version()),
+            "-DCMAKE_BUILD_TYPE={}".format(cfg),  # not used on MSVC, but no harm
+        ]
+        build_args = []
+
+        if self.compiler.compiler_type != "msvc":
+            # Using Ninja-build since it a) is available as a wheel and b)
+            # multithreads automatically. MSVC would require all variables be
+            # exported for Ninja to pick it up, which is a little tricky to do.
+            cmake_args += ["-GNinja"]
 
-        if platform.system() == "Windows":
-            cmake_args += ['-DCMAKE_LIBRARY_OUTPUT_DIRECTORY_{}={}'.format(cfg.upper(), extdir)]
-            if sys.maxsize > 2**32:
-                cmake_args += ['-A', 'x64']
-            build_args += ['--', '/m']
         else:
-            cmake_args += ['-DCMAKE_BUILD_TYPE=' + cfg]
-            build_args += ['--', '-j2']
+            # CMake lets you override the generator - we need to check this.
+            # Can be set with Conda-Build, for example.
+            cmake_generator = os.environ.get("CMAKE_GENERATOR", "")
+
+            # Single config generators are handled "normally"
+            single_config = any(x in cmake_generator for x in {"NMake", "Ninja"})
+
+            # CMake allows an arch-in-generator style for backward compatibility
+            contains_arch = any(x in cmake_generator for x in {"ARM", "Win64"})
+
+            # Specify the arch if using MSVC generator, but only if it doesn't
+            # contain a backward-compatibility arch spec already in the
+            # generator name.
+            if not single_config and not contains_arch:
+                cmake_args += ["-A", PLAT_TO_CMAKE[self.plat_name]]
+
+            # Multi-config generators have a different way to specify configs
+            if not single_config:
+                cmake_args += [
+                    "-DCMAKE_LIBRARY_OUTPUT_DIRECTORY_{}={}".format(cfg.upper(), extdir)
+                ]
+                build_args += ["--config", cfg, "--", "/m"]
```

Comment:
Why don't we add `-j 2` anymore? Maybe `/m` should then also go?

Reply:
Because a) what is special about two cores, and b) Ninja uses all your cores automatically, so it's not needed! MSVC does not do that, so we still add `/m`.

Reply:
Ofc, I read that above, but forgot to put the two pieces of the puzzle together!

Reply:
("what is special about two cores": I think Travis offers two cores, so I'm betting it originated there ;-) )

Reply:
Note: I can't "resolve" things, either. So close at your discretion.

Reply:
As mentioned above, I think it is better if setup.py respects the `-j` command-line option. You can query the option here with `self.parallel`.

Reply:
As mentioned above, "please make sure your computer has make or pip install will fail" is much worse than adding ninja to the build requirements and just using it. And Ninja does this smartly for you. Ninja is also much faster if you have a more complex build, as it builds all sources in parallel, not just by target, which is the best CMake's makefiles can do.

Reply:
Ah, I did forget about the command line option to setup.py, since it's harder to set from pip (I think?). Rereading what you wrote, I see you are refering to that. It also already controls something different in setuptools - it's the number of extensions that build in parallel, rather than the objects in the extension. Unless you have a lot of extensions, it's pretty rarely used. But we could pick it up if set and pass it through to CMake, and set the minimum to 3.12? @wjakob tends to require 3.7 as the minimum, but we are enforcing CMake via pyproject.toml, so that shouldn't be a limitation here?

Reply:
I've attempted to support this, really hoping it is 0 or None by default from looking at IMinuit - I'd expect this to be quite rarely set in practice, since even NumPy came up with a different scheme for multithreading the build. But in that case, Ninja does the right thing, you can control make with MAKEFLAGS, etc. (I've attempted to allow a different generator to be set on UNIX now too).

Reply:
This is Python 3.5+ only, actually, so I'll have to roll it back a bit. And setting it from pip seems to be a pain; is the only way to set it `pip wheel --global-option bdist_ext --global-option  -j4`, which according to [here](https://pip-python3.readthedocs.io/en/latest/reference/pip_wheel.html#customising-the-build) is basically an unsupported accident? I feel like passing arguments into wheel builds is pretty highly frowned upon, which is why environment variables are so popular for this sort of thing.